File: 2018/02/inventory.cc
#include <cassert>
#include <fstream>
#include <iostream>
#include <set>
#include <string>
#include <vector>

typedef struct {
	int	two;
	int	three;
} Inventory;

static void
compute_checksum(Inventory &inventory, std::string boxID)
{
	// This function is void because it really needs to return a tuple -
	// which could be done - and it's single-threaded, so it makes sense to
	// just operate on a reference to an Inventory.

	int	counts[26] = {0};	// assumption: there are only lower-case letters.
	int	two = 0;
	int	three = 0;

	for (char c : boxID) {
		// assumption: letters are lowercase.
		c -= 0x61;
		assert(c >= 0);
		assert(c < 26);

		// Poor man's set.
		counts[static_cast<std::size_t>(c)]++;
	}

	for (auto i = 0; i < 26; i++) {
		if (counts[i] == 2)	two++;
		if (counts[i] == 3)	three++;
	}

	if (two > 0) {
		inventory.two++;
	}
	if (three > 0) {
		inventory.three++;
	}
}

static int
compute_checksums(std::vector<std::string> &boxIDs)
{
	Inventory	inventory = {0, 0};

	for (auto boxID : boxIDs) {
		compute_checksum(inventory, boxID);
	}

	return inventory.two * inventory.three;
}

static void
self_check_checksum(void)
{
	std::vector<std::string>	boxIDs = {
		"abcdef",
		"bababc",
		"abbcde",
		"abcccd",
		"aabcdd",
		"abcdee",
		"ababab",
	};

	auto checksum = compute_checksums(boxIDs);
	assert(checksum == 12);
	std::cout << "self check OK" << std::endl;
}

static std::vector<std::string>
load_inventory(std::string path)
{
	std::ifstream			inventory_file(path);
	std::vector<std::string>	boxIDs;

	while (!inventory_file.eof()) {
		// This could be done in a streaming style, but this approach
		// reuses the function that's been tested in the self test. It
		// also allows reusing the box IDs later on.
		std::string	boxID;
		inventory_file >> boxID;
		boxIDs.push_back(boxID);
	}
	inventory_file.close();
	return boxIDs;
}

typedef struct {
	bool		matched;
	std::string	common;
} MatchResult;

static MatchResult
match_boxes(std::string boxID1, std::string boxID2)
{
	MatchResult	result;
	std::string	common = "";
	int		differences = 0;

	for (std::size_t i = 0; i < boxID1.size(); i++) {
		if (boxID1[i] == boxID2[i]) {
			common.push_back(boxID1[i]);
			continue;
		}

		differences++;
	}

	if (differences == 1) {
		result.matched = true;
		result.common = common;
	}
	else {
		result.matched = false;
	}
	return result;
}

static MatchResult
match_inventory(std::vector<std::string> boxIDs)
{
	MatchResult	result;

	for (std::size_t i = 0; i < boxIDs.size() - 1; i++) {
		for (std::size_t j = i+1; j < boxIDs.size(); j++) {
			result = match_boxes(boxIDs[i], boxIDs[j]);
			if (result.matched) {
				return result;
			}
		}
	}

	return result;
}

static void
self_check_match(void)
{
	std::vector<std::string> boxIDs = {
		"abcde",
		"fghij",
		"klmno",
		"pqrst",
		"fguij",
		"axcye",
		"wvxyz",
	};

	auto result = match_inventory(boxIDs);
	assert(result.matched);
	std::cout << "expected fgij, have " << result.common << std::endl;
	assert(result.common == "fgij");
}

static void
self_check(void)
{
	self_check_checksum();
	self_check_match();
}

int
main(int argc, char *argv[])
{
	self_check();

	for (auto i = 1; i < argc; i++) {
		auto path = std::string(argv[i]);
		auto boxIDs = load_inventory(path);
		
		// Part 1
		auto checksum = compute_checksums(boxIDs);
		std::cout << path << ":\t" << checksum << std::endl;

		// Part 2
		auto result = match_inventory(boxIDs);
		if (result.matched) {
			std::cout << "common: " << result.common << std::endl;
		}
		else {
			std::cout << "no match found" << std::endl;
		}
	}
}

File: advent2018/src/advent/inventory.rs
fn compute_checksum(box_id: String) -> (i64, i64) {
	let mut set: [u8; 26] = [
		0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	];
	let mut two: i64 = 0;
	let mut three: i64 = 0;

	for c in box_id.chars() {
		let i = c as usize;
		set[i - 0x61] += 1;
	}

	for i in 0..set.len() {
		if set[(i as usize)] == 2 {
			two += 1;
		}
		if set[i as usize] == 3 {
			three += 1;
		}
	}

	if two > 0 {
		two = 1;
	}

	if three > 0 {
		three = 1;
	}

	return (two, three);
}

#[test]
fn test_compute_checksum() {
	let (two, three) = compute_checksum("abcdef".to_string());
	assert!(two == 0 && three == 0);

	let (two, three) = compute_checksum("bababc".to_string());
	assert!(two == 1 && three == 1);
}

pub fn compute_checksums(box_ids: &Vec<String>) -> i64 {
	let mut two: i64 = 0;
	let mut three: i64 = 0;

	for box_id in box_ids.iter() {
		let (two_, three_) = compute_checksum(box_id.to_string());
		two += two_;
		three += three_;
	}

	return two * three;
}

#[test]
fn test_compute_checksums() {
	let box_ids: Vec<String> = vec![
		"abcdef".to_string(),
		"bababc".to_string(),
		"abbcde".to_string(),
		"abcccd".to_string(),
		"aabcdd".to_string(),
		"abcdee".to_string(),
		"ababab".to_string(),
	];
	let checksum = compute_checksums(&box_ids);
	assert!(checksum == 12);
}

fn match_boxes(box_id1: String, box_id2: String) -> Option<String> {
	let mut differences = 0;
	let mut common: String = String::new();

	let box_id1_chars = box_id1.as_bytes();
	let box_id2_chars = box_id2.as_bytes();

	for i in 0..box_id1.len() {
		if box_id1_chars[i] == box_id2_chars[i] {
			common.push(box_id1_chars[i] as char);
			continue;
		}

		differences += 1;
		if differences > 1 {
			break;
		}
	}

	if differences == 1 {
		return Some(common);
	}

	None
}

#[test]
fn test_match_boxes() {
	assert!(
		match_boxes("abcde".to_string(), "fghij".to_string()).is_none(),
		"boxes shouldn't match"
	);
	assert!(
		match_boxes("abcde".to_string(), "axcye".to_string()).is_none(),
		"boxes shouldn't match"
	);
	let common = match_boxes("fghij".to_string(), "fguij".to_string());
	match common {
		Some(s) => {
			println!("common: {}", s);
			assert!(s == "fgij".to_string());
		}
		_ => assert!(false),
	};
}

pub fn match_inventory(inventory: &Vec<String>) -> Option<String> {
	for i in 0..(inventory.len() - 1) {
		for j in i + 1..inventory.len() {
			let result = match_boxes(inventory[i].clone(), inventory[j].clone());
			match result {
				Some(common) => return Some(common),
				_ => continue,
			}
		}
	}

	None
}

#[test]
fn test_match_inventory() {
	let inventory = &vec![
		"abcde".to_string(),
		"fghij".to_string(),
		"klmno".to_string(),
		"pqrst".to_string(),
		"fguij".to_string(),
		"axcye".to_string(),
		"wvxyz".to_string(),
	];
	let result = match_inventory(&inventory);
	match result {
		Some(common) => assert!(common == "fgij".to_string()),
		_ => assert!(false),
	};
}
